protocols/identify/src/handler.rs

```diff
@@ -34,89 +47,119 @@ use libp2p_swarm::{
-    pub fn new() -> Self {
+    pub fn new(initial_delay: Duration, interval: Duration) -> Self {
         IdentifyHandler {
-            config: IdentifyProtocolConfig,
             events: SmallVec::new(),
-            next_id: Delay::new(DELAY_TO_FIRST_ID),
+            next_id: Delay::new(initial_delay),
             keep_alive: KeepAlive::Yes,
+            interval,
         }
     }
 }
 
 impl ProtocolsHandler for IdentifyHandler {
-    type InEvent = ();
+    type InEvent = IdentifyPush;
     type OutEvent = IdentifyHandlerEvent;
-    type Error = ReadOneError;
-    type InboundProtocol = IdentifyProtocolConfig;
-    type OutboundProtocol = IdentifyProtocolConfig;
+    type Error = io::Error;
+    type InboundProtocol = SelectUpgrade<IdentifyProtocol, IdentifyPushProtocol<InboundPush>>;
+    type OutboundProtocol = EitherUpgrade<IdentifyProtocol, IdentifyPushProtocol<OutboundPush>>;
     type OutboundOpenInfo = ();
     type InboundOpenInfo = ();
 
     fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
-        SubstreamProtocol::new(self.config.clone(), ())
+        SubstreamProtocol::new(
+            SelectUpgrade::new(
+                IdentifyProtocol,
+                IdentifyPushProtocol::inbound(),
+            ), ())
     }
 
     fn inject_fully_negotiated_inbound(
         &mut self,
-        protocol: <Self::InboundProtocol as InboundUpgrade<NegotiatedSubstream>>::Output,
-        _info: Self::InboundOpenInfo
+        output: <Self::InboundProtocol as InboundUpgrade<NegotiatedSubstream>>::Output,
+        _: Self::InboundOpenInfo
     ) {
-        self.events.push(IdentifyHandlerEvent::Identify(protocol))
+        match output {
+            EitherOutput::First(substream) => {
+                self.events.push(
+                    ProtocolsHandlerEvent::Custom(
+                        IdentifyHandlerEvent::Identify(substream)))
+            }
+            EitherOutput::Second(info) => {
+                self.events.push(
+                    ProtocolsHandlerEvent::Custom(
+                        IdentifyHandlerEvent::Identified(info)))
+            }
+        }
     }
 
     fn inject_fully_negotiated_outbound(
         &mut self,
-        protocol: <Self::OutboundProtocol as OutboundUpgrade<NegotiatedSubstream>>::Output,
-        _info: Self::OutboundOpenInfo,
+        output: <Self::OutboundProtocol as OutboundUpgrade<NegotiatedSubstream>>::Output,
+        _: Self::OutboundOpenInfo,
     ) {
-        self.events.push(IdentifyHandlerEvent::Identified(protocol));
-        self.keep_alive = KeepAlive::No;
+        match output {
+            EitherOutput::First(remote_info) => {
+                self.events.push(
+                    ProtocolsHandlerEvent::Custom(
+                        IdentifyHandlerEvent::Identified(remote_info)));
+                self.keep_alive = KeepAlive::No;
+            }
+            EitherOutput::Second(()) => {}
```

Comment:
Something that came to my mind: How about emitting an event along the lines of `IdentifyEvent::Pushed` once a push succeeded? I would expect this to be useful for people tracking these events in a monitoring system like Prometheus. This event would serve a similar purpose to the `RequestSent` event in `libp2p-request-response`:

https://github.com/libp2p/rust-libp2p/blob/63512e5f160c058471e59e9e7c48a40cb503c1d0/protocols/request-response/src/lib.rs#L152-L161

In case you think this is useful @romanb I can prepare a pull request.

Reply:
Sounds good to me.